dali/test/python/test_operator_cast.py

```diff
@@ -12,46 +12,106 @@
-    eii2 = RandomlyShapedDataIterator(batch_size, max_shape=input_shape, dtype=in_type)
-    compare_pipelines(
-        CastPipeline(device, batch_size, iter(eii1), [types.FLOAT16, out_type]),
-        CastPipeline(device, batch_size, iter(eii2), [out_type]),
-        batch_size=batch_size, N_iterations=5)
-
-def test_cast_operator_float16():
+from test_utils import check_batch, np_type_to_dali
+
+
+def ref_cast(x, dtype):
+    if np.issubdtype(dtype, np.integer):
+        lo = np.iinfo(dtype).min
+        hi = np.iinfo(dtype).max
+        if np.issubdtype(x.dtype, np.floating):
+            x = np.round(x)
+        else:
+            x
+        return x.clip(lo, hi).astype(dtype)
+    else:
+        return x.astype(dtype)
+
+def random_shape(rng, ndim:int, max_size:int):
+    if ndim == 0:
+        return []
+    max_size = int(max_size ** (1/ndim))
+    return list(rng.integers(0, max_size, [ndim]))
+
+def generate(rng, ndim:int, batch_size:int, in_dtype:np.dtype, out_dtype:np.dtype):
+    lo, hi = -1000, 1000
+    if np.issubdtype(out_dtype, np.integer):
+        lo = np.iinfo(out_dtype).min
+        hi = np.iinfo(out_dtype).max
+        if hi < np.iinfo(np.int64).max:
+            r = hi - lo
+            hi += r // 2
+            lo -= r // 2
+        if np.issubdtype(in_dtype, np.integer):
+            lo = max(np.iinfo(in_dtype).min, lo)
+            hi = min(np.iinfo(in_dtype).max, hi)
+        else:
+            lo = max(-np.finfo(in_dtype).max, lo)
+            hi = min(np.finfo(in_dtype).max, hi)
+
+    max_size = 100000 // batch_size
+    out = [rng.uniform(lo, hi, size=random_shape(rng, ndim, max_size)).astype(in_dtype) for _ in range(batch_size)]
+    if np.issubdtype(in_dtype, np.floating) and np.issubdtype(out_dtype, np.integer):
+        for x in out:
+            # avoid exactly halfway numbers - rounding is different for CPU and GPU
+            halfway = x[x - np.floor(x) == 0.5]
+            x[x - np.floor(x) == 0.5] = np.nextafter(halfway, np.Infinity)
+    return out
+
+rng = np.random.default_rng(1234)
+
+@nottest
+def _test_operator_cast(ndim, batch_size, in_dtype, out_dtype, device):
+    src = lambda: generate(rng, ndim, batch_size, in_dtype, out_dtype)
+    @pipeline_def(batch_size=batch_size, num_threads=4, device_id=types.CPU_ONLY_DEVICE_ID if device == 'cpu' else 0)
+    def cast_pipe():
+        inp = fn.external_source(src)
+        inp_dev = inp.gpu() if device == 'gpu' else inp
+        return inp, fn.cast(inp_dev, dtype=np_type_to_dali(out_dtype))
+
+    pipe = cast_pipe()
+    pipe.build()
+    for _ in range(10):
+        inp, out = pipe.run()
+        if device=='gpu':
+            out = out.as_cpu()
+        ref = [ref_cast(np.array(x), out_dtype) for x in inp]
+
+        # workaround a bug in numpy - it can't make a float16 scalar - it is promoted to double, somehow :(
+        eps = 0 if np.issubdtype(out_dtype, np.integer) else (np.nextafter(out_dtype([1]), 2) - 1.0)[0]
```

Comment:
```suggestion
        if np.issubdtype(out_dtype, np.integer):
            eps = 0
        else:
            # workaround a bug in numpy - it can't make a float16 scalar - it is promoted to double, somehow :(
            eps = (np.nextafter(out_dtype([1]), 2) - 1.0)[0]
```

Just to make it more clear what is worked around here - computing the proper eps and not setting some bigger eps for instance because of some float16 problems. NB, it's not just float16, float32 scalars are promoted too.

Reply:
Meh, nextafter doens't work anyway.